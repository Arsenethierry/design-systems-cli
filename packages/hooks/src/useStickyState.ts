/* eslint-disable no-cond-assign, no-continue */

import { RefObject, useEffect, useState } from "react";

import { debounce } from "@design-systems/utils";

/** Get the nearest scroll container */
function getScrollParent(element: HTMLElement) {
  let style = getComputedStyle(element);

  const excludeStaticParent = style.position === "absolute";
  const overflowRegex = /(auto|scroll|hidden)/;

  if (style.position === "fixed") {
    return document.body;
  }

  let parent: HTMLElement | null = element;

  while ((parent = parent.parentElement)) {
    style = getComputedStyle(parent);

    if (excludeStaticParent && style.position === "static") {
      continue;
    }

    if (
      overflowRegex.test(style.overflow + style.overflowY + style.overflowX)
    ) {
      return parent;
    }
  }

  return document.body;
}

/**
 * Determine if a position:sticky element is sticking.
 *
 * @param element - The element to detect the sticky state of
 *
 * @example
 * const ref = useRef<HTMLDivElement>(null);
 * const isSticking = useStickyState(ref);
 */
export const useStickyState = <T extends HTMLElement>(
  element: RefObject<T | null>
) => {
  const [isSticky, setIsSticky] = useState(false);

  useEffect(() => {
    const { current: stickyTarget } = element;

    if (!stickyTarget || !window) {
      return;
    }

    const parentElement = getScrollParent(stickyTarget);

    if (!parentElement) {
      return;
    }

    const calculateIsSticky = debounce(() => {
      const pTop = parentElement.scrollTop;
      const tTop = stickyTarget.getBoundingClientRect().top;
      const difference = Math.round(Math.abs(pTop - tTop));
      const triggerDistance = window
        .getComputedStyle(stickyTarget)
        .getPropertyValue("top");

      setIsSticky(difference !== Math.abs(parseInt(triggerDistance, 10)));
    }, 25);

    calculateIsSticky();

    parentElement.addEventListener("scroll", calculateIsSticky);

    return () => parentElement.removeEventListener("scroll", calculateIsSticky);
  }, [element]);

  return isSticky;
};