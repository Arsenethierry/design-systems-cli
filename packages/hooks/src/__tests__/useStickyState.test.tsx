import * as React from 'react';
import { render, act } from '@testing-library/react';
import { useStickyState } from '../useStickyState';

test('it should detect when an element is not sticking', () => {
  jest.useFakeTimers();
  const Test = () => {
    const ref = React.useRef<HTMLDivElement>(null);
    const isSticky = useStickyState(ref);

    return (
      <div
        ref={ref}
        data-testid="target"
        style={{ position: 'sticky', top: 0 }}
      >
        {isSticky ? 'sticky' : 'not sticky'}
      </div>
    );
  };

  const { container } = render(
    <div data-testid="root" style={{ height: 1000 }}>
      <Test />
    </div>
  );

  act(() => {
    jest.runOnlyPendingTimers();
  });

  expect(container.firstChild).toMatchSnapshot();
});

test('it should detect when an element is sticking', () => {
  jest.useFakeTimers();
  const Test = () => {
    const ref = React.useRef<HTMLDivElement>(null);
    const isSticky = useStickyState(ref);

    return (
      <div
        ref={ref}
        data-testid="target"
        style={{ position: 'sticky', top: 10 }}
      >
        {isSticky ? 'sticky' : 'not sticky'}
      </div>
    );
  };

  const { container } = render(
    <div data-testid="root" style={{ height: 1000 }}>
      <Test />
    </div>
  );

  act(() => {
    jest.runOnlyPendingTimers();
  });

  expect(container.firstChild).toMatchSnapshot();
});
