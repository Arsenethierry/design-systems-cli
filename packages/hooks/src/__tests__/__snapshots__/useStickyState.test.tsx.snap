// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should detect when an element is not sticking 1`] = `
<div
  data-testid="root"
  style="height: 1000px;"
>
  <div
    data-testid="target"
    style="position: sticky; top: 0px;"
  >
    not sticky
  </div>
</div>
`;

exports[`it should detect when an element is sticking 1`] = `
<div
  data-testid="root"
  style="height: 1000px;"
>
  <div
    data-testid="target"
    style="position: sticky; top: 10px;"
  >
    sticky
  </div>
</div>
`;
